Support UltraRISC DP1000

WangJia-UR opened the pull request:
Added support for UltraRISC DP1000 SoC.
Build and boot testing passed on UR-DP1000-EVB

Related issues: #71

sterling-teng commented on drivers/pci/controller/dwc/pcie-designware-host.c:
```diff
@@ -119,6 +122,18 @@ static void dw_pci_setup_msi_msg(struct irq_data *d, struct msi_msg *msg)
+
+	if (data->chip->irq_set_affinity)
+		return data->chip->irq_set_affinity(data, mask, force);
+#endif
```
这种情况，包括类似的，显得有点把原函数弄乱了，建议单独构造一个函数，然后在调用的地方加编译条件，最后建议添加几行简洁的注释。

iamwjia replied:
OK，我会将dwc中公共代码修改部分都按这个要求调整。

WangJia-UR replied:
OK，我会把dwc中公共代码修改部分都按这个要求调整。

sterling-teng commented on drivers/pci/controller/dwc/pcie-designware.h:
```diff
@@ -90,6 +93,15 @@
+#define PORT_FLT_SF_1024    PORT_FLT_SF(0x0)
+#define PORT_FLT_SF_256     PORT_FLT_SF(0x1)
+#define PORT_FLT_SF_64      PORT_FLT_SF(0x2)
+#define PORT_FLT_SF_16      PORT_FLT_SF(0x3)
```
这个头文件中新增的这些定义在synopsys规范中有写吗？如果有，请适当加些注释，如果没有请移到.c文件顶部。

WangJia-UR replied:
这是规范中的，我会补充注释

sterling-teng commented on arch/riscv/Kconfig.socs:
```diff
@@ -70,6 +70,12 @@ config ARCH_XUANTIE
 	help
```
补丁的描述信息里建议写下soc的规格以及各项指标（就是介绍下soc）。

WangJia-UR replied:
OK，我会补充到commit message

sterling-teng commented on drivers/pci/controller/dwc/pcie-designware-host.c:
```diff
@@ -116,11 +119,41 @@ static void dw_pci_setup_msi_msg(struct irq_data *d, struct msi_msg *msg)
+ */
+static int dw_pci_msi_set_chip_affinity(struct irq_data *d,
+				   const struct cpumask *mask, bool force)
+{
```
你已经在调用处加编译条件了，这里就不需要加了，只需要吧函数名字改为与soc强相关的，比如dw_pci_xxx_dp1000()

WangJia-UR replied:
OK，已增加dp1000后缀

sterling-teng commented on drivers/pci/controller/dwc/pcie-designware-host.c:
```diff
@@ -677,7 +677,7 @@ EXPORT_SYMBOL_GPL(dw_pcie_own_conf_map_bus);
 	.map_bus = dw_pcie_own_conf_map_bus,
-#ifdef CONFIG_PCIE_ULTRARISC
+#if IS_ENABLED(CONFIG_PCIE_ULTRARISC)
 	.read = pci_generic_config_read32,
```
这一处修改，建议合并到上一个补丁中。

WangJia-UR replied:
OK，已修改

sterling-teng commented on drivers/pci/controller/dwc/pcie-designware-host.c:
```diff
@@ -686,6 +686,46 @@ static struct pci_ops dw_pcie_ops = {
+ *
+ * Returns 0 on success, or a negative error code on failure.
+ */
+static int dw_pcie_prog_outbound_atu_multi(struct dw_pcie *pci, struct resource_entry *entry, int *index)
```
如果是synopsys规范中有写，则保持不动。如果没有写，建议函数名改为与soc强相关的名字。

WangJia-UR replied:
已增加dp1000后缀

sterling-teng commented on drivers/net/ethernet/stmicro/stmmac/Kconfig:
```diff
@@ -266,6 +266,13 @@ config DWMAC_VISCONTI
 	help
 	  Support for ethernet controller on Visconti SoCs.
 
+config DWMAC_ULTRARISC
```
请补充补丁的描述信息，至少要包括mac的规格指标。

WangJia-UR replied:
已补充到commit message

sterling-teng commented on drivers/pinctrl/ultrarisc/Makefile:
```diff
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier: GPL-2.0-only
+
+obj-$(CONFIG_PINCTRL_ULTRARISC) += pinctrl-ultrarisc.o
+obj-$(CONFIG_PINCTRL_ULTRARISC_DP1000)	+= pinctrl-ultrarisc-dp1000.o
```
文件结尾空一行。

WangJia-UR replied:
已修改

sterling-teng commented on drivers/pinctrl/ultrarisc/pinctrl-ultrarisc.c:
```diff
@@ -0,0 +1,501 @@
+			unsigned long *config)
+{
+	dev_dbg(pctldev->dev, "%s(%d): pin=%d\n", __func__, __LINE__, pin);
+	// TODO: this is call by pinconf-generic
```
似乎没有完成？我不反对这行注释，是否保留你决定。

WangJia-UR replied:
已移除

sterling-teng commented on drivers/pinctrl/ultrarisc/pinctrl-ultrarisc.h:
```diff
@@ -0,0 +1,78 @@
+	const struct pinctrl_pin_desc *pins;
+	u32 npins;
+	u32 offset;
+	//u32 conf_offset[];
```
drop it?

WangJia-UR replied:
类似的都已移除

sterling-teng commented on Documentation/devicetree/bindings/pinctrl/ultrarisc,dp1000-pinctrl.yaml:
```diff
@@ -0,0 +1,105 @@
+          UR_DP1000_BIAS(UR_DP1000_IOMUX_B, 7, UR_PULL_UP, UR_DRIVE_DEF)
+          >;
+      };
+    };
```
末尾留一个空行。

WangJia-UR replied:
OK

sterling-teng commented on arch/riscv/configs/dp1000_defconfig:
```diff
@@ -0,0 +1,5530 @@
+CONFIG_LOCALVERSION="-ur-dp1000-rvck"
```
我们现在使用defconfig，但是挡墙的defconfig还处于调试开发状态，预计周三之前会完成，届时请另行适配defconfig。

WangJia-UR replied:
OK

sterling-teng commented on arch/riscv/boot/dts/ultrarisc/dp1000.dts:
```diff
@@ -467,7 +467,7 @@
 			ranges = <0x81000000  0x0 0x4fbf0000  0x0 0x4fbf0000  0x0 0x00400000>, /* io */
 				 <0x82000000  0x0 0x40000000  0x0 0x40000000  0x0 0x0fbf0000>, /* mem32 */
-				 <0xc3000000 0x40 0x00000000 0x40 0x00000000  0x5 0x00000000>; /* mem64 prefetchable */
+				 <0xc3000000 0x40 0x00000000 0x40 0x00000000  0xd 0x00000000>; /* mem64 prefetchable */
```
注意到前面的补丁有涉及到“16 out windows”，该补丁的编号是否需要调整？是否适用二分法？

WangJia-UR replied:
适用二分法，但我还是调整了顺序

sterling-teng commented on drivers/pci/msi/msi.c:
```diff
@@ -12,7 +12,9 @@
 #include "msi.h"
+#ifdef CONFIG_ARCH_SOPHGO
 #include "../controller/cadence/pcie-cadence-sophgo.h"
+#endif
```
这个补丁暂时合入，但是后期可能会剔除，因为已经有人提交了sg2042的上游方案，目前还在调研中。

WangJia-UR replied:
OK

sterling-teng commented on arch/riscv/mm/init.c:
```diff
@@ -1455,7 +1455,7 @@ static void __init create_linear_mapping_page_table(void)
 			start = __pa(PAGE_OFFSET);
 
-		create_linear_mapping_range(start, end, PMD_SIZE);
+		create_linear_mapping_range(start, end, 0);
```
该补丁与74号pr部分改动一致，因为你的pr提交时间更早，原则上优先合并你的，但是74号pr比较急，所以我今天会暂时合入74号pr的方案，得益于开发分支的滚动方式，我可以在合并后进行调整，如果你希望合并这个补丁，请尽快回复。

WangJia-UR replied:
如果可以的话，还请帮我合入

sterling-teng commented:
补丁已经审阅完了，总体上，补丁的代码部分基本符合规范，描述信息还需要参考规范（规范并不明朗，近期会完善）修改一下，建议参考74号pr的补丁。

我在尝试应用补丁的时候，git输出了一下警告：

```
Applying: riscv:pcie: support dw pcie rc interrupt affinity settings in DP1000
.git/rebase-apply/patch:48: trailing whitespace.

warning: 1 line adds whitespace errors.
Applying: riscv:PCI: Update dw_pcie_ops for 32-bit cfg access in DP1000
Applying: pcie: update the outbound mapping process
Applying: net:stmmac: add gmac driver for UltraRISC DP1000
Applying: pinctrl: add pinctrl dirver of UltraRISC DP1000
Applying: dt-bindings:pinctrl: add pincltrl binding file of DP1000
.git/rebase-apply/patch:29: trailing whitespace.
  
.git/rebase-apply/patch:32: trailing whitespace.
  
.git/rebase-apply/patch:66: trailing whitespace.
              node apply to. This should be set using the UR_DP1000_IOPAD 
.git/rebase-apply/patch:98: trailing whitespace.
          
.git/rebase-apply/patch:110: trailing whitespace.
          
warning: 5 lines add whitespace errors.

```
请修复。

另外我在执行checkpatch.pl时，也有错误和警告输出，请修复。

```
WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
```

该行警告可忽略，其它的如果在修复中遇到问题，及时讨论。

WangJia-UR commented:
> 补丁已经审阅完了，总体上，补丁的代码部分基本符合规范，描述信息还需要参考规范（规范并不明朗，近期会完善）修改一下，建议参考74号pr的补丁。
> 
> 我在尝试应用补丁的时候，git输出了一下警告：
> 
> ```
> Applying: riscv:pcie: support dw pcie rc interrupt affinity settings in DP1000
> .git/rebase-apply/patch:48: trailing whitespace.
> 
> warning: 1 line adds whitespace errors.
> Applying: riscv:PCI: Update dw_pcie_ops for 32-bit cfg access in DP1000
> Applying: pcie: update the outbound mapping process
> Applying: net:stmmac: add gmac driver for UltraRISC DP1000
> Applying: pinctrl: add pinctrl dirver of UltraRISC DP1000
> Applying: dt-bindings:pinctrl: add pincltrl binding file of DP1000
> .git/rebase-apply/patch:29: trailing whitespace.
>   
> .git/rebase-apply/patch:32: trailing whitespace.
>   
> .git/rebase-apply/patch:66: trailing whitespace.
>               node apply to. This should be set using the UR_DP1000_IOPAD 
> .git/rebase-apply/patch:98: trailing whitespace.
>           
> .git/rebase-apply/patch:110: trailing whitespace.
>           
> warning: 5 lines add whitespace errors.
> ```
> 
> 请修复。
> 
> 另外我在执行checkpatch.pl时，也有错误和警告输出，请修复。
> 
> ```
> WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
> ```
> 
> 该行警告可忽略，其它的如果在修复中遇到问题，及时讨论。

已使用checkpatch.pl扫描并完成清理

sterling-teng commented:
如果一切顺利，预计本周四晚上出基为6.6.100或6.6.101的tag，如果遇到问题，可能延迟到下周一之前。制作tag期间不会合入补丁，但是正常进行审阅。请把握好更新pr的时间。

sterling-teng commented:
该pr已经合并，共19个补丁，应用了18个。

1 这三行开头有一个空格，应用时已删除。

```
+ 	  UR-DP1000 is an 8-core 64-bit RISC-V SoC that supports
+ 	  the RV64GCBHX ISA. Supports Hardware Virtualization,
+ 	  RISC-V RV64 ISA H(v1.0) Extension.
```

2 去掉最后一个补丁

经过评估，riscv: mm: Adjust fixed_map_size in create_linear_mapping_range是Revert "riscv: kexec: Add image loader for kexec file"的一部分，拆分后会导致git log不清晰，所以去掉该补丁。

WangJia-UR commented:
> 该pr已经合并，共19个补丁，应用了18个。
> 
> 1 这三行开头有一个空格，应用时已删除。
> 
> ```
> + 	  UR-DP1000 is an 8-core 64-bit RISC-V SoC that supports
> + 	  the RV64GCBHX ISA. Supports Hardware Virtualization,
> + 	  RISC-V RV64 ISA H(v1.0) Extension.
> ```
> 
> 2 去掉最后一个补丁
> 
> 经过评估，riscv: mm: Adjust fixed_map_size in create_linear_mapping_range是Revert "riscv: kexec: Add image loader for kexec file"的一部分，拆分后会导致git log不清晰，所以去掉该补丁。

OK, thx
